lms/djangoapps/certificates/badge_handler.py

```diff
@@ -0,0 +1,134 @@
+
+    def course_slug(self, mode):
+        """
+        Slug ought to be deterministic and limited in size so it's not too big for Badgr.
+        """
+        return hashlib.sha256(u"{}{}".format(unicode(self.course_key), unicode(mode))).hexdigest()
+
+    def log_if_raised(self, response):
+        """
+        Log server response if there was an error.
+        """
+        try:
+            response.raise_for_status()
+        except HTTPError:
+            LOGGER.error(
+                u"Encountered an error when contacting the Badgr-Server. Response status was %s.\n%s",
+                response.status_code, response.body
+            )
+            raise
+
+    def get_headers(self):
+        """
+        Headers to send along with the request-- used for authentication.
+        """
+        return {'Authorization': 'Token {}'.format(settings.BADGR_API_TOKEN)}
+
+    def ensure_badge_created(self, mode):
+        """
+        Verify a badge has been created for this mode of the course, and, if not, create it
+        """
+        if self.course_slug(mode) in BadgeHandler.badges:
+            return
+        response = requests.get(self.badge_url(mode), headers=self.get_headers())
+        if response.status_code != 200:
+            self.create_badge(mode)
+        BadgeHandler.badges[self.course_slug(mode)] = True
+
+    def create_badge(self, mode):
+        """
+        Create the badge spec for a course's mode.
+        """
+        course = modulestore().get_course(self.course_key)
+        if mode not in settings.BADGR_IMAGE_SOURCES:
+            raise KeyError("No image specified for course mode {} in BADGR_IMAGE_SOURCES!".format(mode))
+        image_name = settings.BADGR_IMAGE_SOURCES[mode]
+        image = open(image_name, 'rb')
+        # This will fail if the filename is something silly.
+        content_type = 'image/{}'.format(image_name.rsplit('.', 1)[-1].lower())
+        files = {'image': (image_name, image, content_type)}
+        try:
+            domain = unicode(Site.objects.get_current().domain)
+        except Site.DoesNotExist:
+            domain = u'example.com'
+        about_path = reverse('about_course', kwargs={'course_id': unicode(self.course_key)})
+        data = {
+            'name': course.display_name,
+            'criteria': u'http://{}{}'.format(domain, about_path),
+            'slug': self.course_slug(mode)
+        }
+        result = requests.post(self.badge_create_url, headers=self.get_headers(), data=data, files=files)
+        self.log_if_raised(result)
+
+    def create_assertion(self, user, mode):
+        """
+        Register an assertion with the Badgr server for a particular user in a particular course mode for
+        this course.
+        """
+        response = requests.post(self.assertion_url(mode), headers=self.get_headers(), data={'email': user.email})
+        self.log_if_raised(response)
+        BadgeTracker(course_id=self.course_key, user=user, data=response.json()).save()
```

Comment:
@smarnach I suppose if someone ran the regeneration command quickly twice in a row, this could happen. I'll code a contingency for it.